adapters/Vungle/VungleAdapter/Bidding/GADMediationVungleNativeAd.m

```diff
@@ -0,0 +1,248 @@
+        delegate = originalCompletionHandler(ad, error);
+      }
+      originalCompletionHandler = nil;
+      return delegate;
+    };
+  }
+  return self;
+}
+
+- (void)requestAd {
+  self.desiredPlacement = [GADMAdapterVungleUtils findPlacement:_adConfiguration.credentials.settings
+                                                  networkExtras:_adConfiguration.extras];
+  if (!self.desiredPlacement) {
+    NSError *error = GADMAdapterVungleErrorWithCodeAndDescription(GADMAdapterVungleErrorInvalidServerParameters, @"Placement ID not specified.");
+    _adLoadCompletionHandler(nil, error);
+    return;
+  }
+    
+  if (![[VungleSDK sharedSDK] isInitialized]) {
+    NSString *appID = [GADMAdapterVungleUtils findAppID:_adConfiguration.credentials.settings];
+    [[GADMAdapterVungleRouter sharedInstance] initWithAppId:appID delegate:self];
+    return;
+  }
+    
+  [self loadAd];
+}
+
+- (void)loadAd {
+  _nativeAd = [[VungleNativeAd alloc] initWithPlacementID:self.desiredPlacement];
+  _nativeAd.delegate = self;
+  VungleAdNetworkExtras *networkExtras = [_adConfiguration extras];
+  if (networkExtras.nativeAdOptionPosition != 0) {
+    _nativeAd.adOptionsPosition = networkExtras.nativeAdOptionPosition;
+  }
+  [_nativeAd loadAd];
+}
+
+#pragma mark - GADMediatedUnifiedNativeAd
+
+- (nullable NSString *)headline {
+    return _nativeAd.title;
+}
+
+- (nullable NSArray<GADNativeAdImage *> *)images {
+  return nil;
+}
+
+- (nullable NSString *)body {
+    return _nativeAd.bodyText;
+}
+
+- (nullable GADNativeAdImage *)icon {
+    return [[GADNativeAdImage alloc] initWithImage:_nativeAd.iconImage];
+}
+
+- (nullable NSString *)callToAction {
+    return _nativeAd.callToAction;
+}
+
+- (nullable NSDecimalNumber *)starRating {
+  return [[NSDecimalNumber alloc] initWithDouble:_nativeAd.adStarRating];
+}
+
+- (nullable NSString *)store {
+  return nil;
+}
+
+- (nullable NSString *)price {
+  return nil;
+}
+
+- (nullable NSString *)advertiser {
+  return _nativeAd.sponsoredText;
+}
+
+- (nullable NSDictionary<NSString *, id> *)extraAssets {
+  return nil;
+}
+
+- (nullable UIView *)adChoicesView {
+  return nil;
+}
+
+- (nullable UIView *)mediaView {
+  return _mediaView;
+}
+
+- (BOOL)hasVideoContent {
+  return YES;
+}
+
+- (void)didRenderInView:(nonnull UIView *)view
+    clickableAssetViews:(nonnull NSDictionary<GADNativeAssetIdentifier, UIView *> *)clickableAssetViews
+ nonclickableAssetViews:(nonnull NSDictionary<GADNativeAssetIdentifier, UIView *> *)nonclickableAssetViews
+         viewController:(nonnull UIViewController *)viewController {
+  NSArray<UIView *> *assets = clickableAssetViews.allValues;
+  UIImageView *iconView = nil;
+  if ([clickableAssetViews[GADNativeIconAsset] isKindOfClass:[UIImageView class]]) {
+    iconView = (UIImageView *)clickableAssetViews[GADNativeIconAsset];
+  }
+    
+  [_nativeAd registerViewForInteraction:view mediaView:_mediaView iconImageView:iconView viewController:viewController clickableViews:assets];
```

Comment:
Would you mind clarifying how the Vungle SDK tracks clicks and impressions?

Reply:
Impression callback is fired when the ad is rendered. The click callback is fired when the user taps on the ad. The callbacks are handled in this ad class in the `VungleNativeAdDelegate` section below. The class is registered as the native ad delegate at L91.